### ISSUE TITLE
[ParamountPlus] Show extractor doesn't fetch all episodes...

### ISSUE THREAD

**Issue by spookyahell** [yt-dlp/yt-dlp#3155]
### Checklist

- [X] I'm reporting a broken site
- [X] I've verified that I'm running yt-dlp version **2022.03.08.1**. ([update instructions](https://github.com/yt-dlp/yt-dlp#update))
- [X] I've checked that all provided URLs are alive and playable in a browser
- [X] I've checked that all URLs and arguments with special characters are [properly quoted or escaped](https://github.com/ytdl-org/youtube-dl#video-url-contains-an-ampersand-and-im-getting-some-strange-output-1-2839-or-v-is-not-recognized-as-an-internal-or-external-command)
- [X] I've searched the [bugtracker](https://github.com/yt-dlp/yt-dlp/issues?q=) for similar issues including closed ones. DO NOT post duplicates
- [X] I've read the [guidelines for opening an issue](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#opening-an-issue)
- [X] I've read about [sharing account credentials](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#are-you-willing-to-share-account-details-if-needed) and I'm willing to share it if required

### Region

United States

### Description

...apparently the maximum length of episode list is exceeded

Example show: [Link to show](https://www.paramountplus.com/shows/house-of-anubis/)

As mentioned in #602, yt-dlp only returns 50 episodes for the above show example, but show has 112 episodes as properly returned by the json info, my fix below (patch for paramountplus.py) works and I encourage any of the maintainers to apply my fix.

But... any other solution (e.g. a more elegant one) would also be appreciated as well...

```
   def _entries(self, show_name):
        show_json = self._download_json(self._API_URL.format(show_name, 0), video_id=show_name)
        if show_json.get('success'):
            total = show_json['result']['total']
            episodes = show_json['result']['data']
            x, y = divmod(total, 50)
            if y != 0:
                x += 1 # x is the amount of times (total amount; we already have done it once at this point) we need to fetch the API url with additional pages
            if x > 1: # only if x is larger than one (=the fetch already done above), we will need additional fetches
                for i in range(1,x):
                    show_json = self._download_json(self._API_URL.format(show_name, i), video_id=show_name)
                    if show_json.get('success'):
                        episodes.extend(show_json['result']['data'])
            for episode in episodes:
                yield self.url_result(
                    'https://www.paramountplus.com%s' % episode['url'],
                    ie=ParamountPlusIE.ie_key(), video_id=episode['content_id'])
```

### Verbose log

```shell
DOES NOT APPLY (the problem is described above)
Essentially, I can't providing missing output for the missing episodes, for this reason:
It's missing.
```

**Comment by pukkandan** [comment]
> DOES NOT APPLY (the problem is described above)
Essentially, I can't providing missing output for the missing episodes, for this reason:
It's missing.

It DOES APPLY! You should have put the log of `yt-dlp -vU URL` (you could also add --flat-playlist/--simulate to speed it up) to demonstrate that not all episodes are being downloaded

But I understand what is happening here even without the log, so I'll let this pass

**Comment by dirkf** [comment]
We seem to be collecting episodes from all seasons (season 0) with
https://www.paramountplus.com/shows/{show_name}/xhr/episodes/page/{page}/size/50/xs/0/season/0

Doesn't iterating on `page` fetch all the episodes in chunks of 50?

(404 in UK)

**Comment by spookyahell** [comment]
Yes, but... to "iterate" over page you have to know how many pages exist

(Unless of course you're ok with requesting either a certain number of pages maximum (e.g. via range) or in a while-loop... untill finally some page returns empty results)

The code example I provided above does it with calculation of needed pages.

_Heres a simplified flow chart of what happens_
Fetch first page —> Calculate how many additional pages are necessary, if any —>  Fetch additional pages
if AmountOfEpisodes<51: **don't fetch anything else**
(always at the end) —> transfer info of fetched episodes

PS: All of this is still largly flawed for lack of method to download individual seasons...

**Comment by dirkf** [comment]
> ...until finally some page returns empty results

I believe that's what happens now, quitting when the JSON doesn't give `success` (also crashing if the URL gives 404, etc):
```py
    def _entries(self, show_name):
        for page in itertools.count():
            show_json = self._download_json(
                f'https://www.paramountplus.com/shows/{show_name}/xhr/episodes/page/{page}/size/50/xs/0/season/0', show_name)
            if not show_json.get('success'):
                return
            for episode in show_json['result']['data']:
                yield self.url_result(
                    'https://www.paramountplus.com%s' % episode['url'],
                    ie=ParamountPlusIE.ie_key(), video_id=episode['content_id'])
```
Apparently, to get by season, use `.../season/{season_num}` instead.

Maybe 50 isn't the ideal chunking size, as it just fails to get a year of weekly shows.

**Comment by spookyahell** [comment]
I reported this while unaware that the issue I was reporting was probably already fixed to some extent.
I had not been aware of ec47c12 (which is a commit that has been made since the time of the last release).

I still think my above mentioned approach is somewhat better:
It causes considerably less empty page requests on the endpoint of the service.

@dirkf I think I know what you mean. -
Is [this an example](https://www.paramountplus.com/shows/the-late-show-with-stephen-colbert/) where it doesn't work or could you give me one example where that is the case?

**Comment by pukkandan** [comment]
Can you clarify whether https://github.com/yt-dlp/yt-dlp/commit/ec47c12f69857f09a79787a7346c957f5b3d4e16 fetches all the pages?

Since you were making code suggestions, I assumed you are on the git master. I really should have forced you to show the log 😔

If it does, your solution is not more "elegant".
```
x, y = divmod(total, 50)
```
making this assumption only makes the code more likely to break in the future. And the number of API requests made are the same using either method.

**Comment by spookyahell** [comment]
I'm sorry about rhe confusion with what version of YouTube DL I was using.
I thought that it was clear when I put the version number that I was in rhat release.
And not the latest git. I will now know this for the future and I will further clarify when I find it necessary.

However, your statement about the existing code not making more requests is wrong.

The function currently has a return statement to be executed on API failure. ("when not success")
This happens on a non-existing page.
That ist not a good solution when it is possible to do it one better.

Sure things can break, I don't see how this one is more subject to breaking than the other solution
